### PR TITLE
Fixed #27995 -- Added error messages on unsupported operations following union(), intersection(), and difference().

### PR DESCRIPTION
**Pull request by hramezani**
[Ticket 27995](https://code.djangoproject.com/ticket/27995)

### CODE REVIEW

File: django/db/models/query.py [hunk 1]
```diff
@@ -904,6 +904,9 @@ def _filter_or_exclude(self, negate, *args, **kwargs):
             assert self.query.can_filter(), \
                 "Cannot filter a query once a slice has been taken."
 
+            if self.query.combinator in ['union', 'intersection', 'difference']:
```
Comment (charettes) [inline django/db/models/query.py]: I think we want to fail as soon as `combinator` is truthy

```suggestion
            if self.query.combinator:
```

File: django/db/models/query.py [hunk 2]
```diff
@@ -904,6 +904,9 @@ def _filter_or_exclude(self, negate, *args, **kwargs):
...
+                raise TypeError('Cannot call .%s() after %s().' %
+                                (['filter', 'exclude'][negate], self.query.combinator))
```
Comment (charettes) [inline django/db/models/query.py]: Clever use of the integer nature of `bool`, first time I've seen this pattern. It kind of hurts readability though IMO.

File: django/db/models/query.py [hunk 3]
```diff
@@ -904,6 +904,9 @@ def _filter_or_exclude(self, negate, *args, **kwargs):
             assert self.query.can_filter(), \
                 "Cannot filter a query once a slice has been taken."
 
+            if self.query.combinator:
+                raise TypeError('Cannot call .%s() after %s().' %
```
Comment (felixxm) [inline django/db/models/query.py]: I would use `NotSupportedError`.

**Comment by hramezani** [comment]
@charettes Thanks. Requested changes are done.

File: django/db/models/query.py [hunk 4]
```diff
@@ -904,6 +904,9 @@ def _filter_or_exclude(self, negate, *args, **kwargs):
...
+                                ('exclude' if negate else 'filter', self.query.combinator))
```
Comment (felixxm) [inline django/db/models/query.py]: Maybe e.g. `Calling exclude() is not supported after union().`

**Comment by hramezani** [comment]
@felixxm , Thanks for the comments. I added `_supported_operation` decorator for checking the supported operation after `union()`, `intersection()`, and `difference()`, also applied it to `filter(), exclude(), defer(), only(), prefetch_related(), select_related(), and extra()`.
Some tests did not pass because they call `.get()` after the combinator and `.get()` also call `.filter()`.
https://github.com/django/django/blob/fc75694257b5bceab82713f84fe5a1b23d641c3f/tests/queries/test_qs_combinators.py#L84

Should we raise an error for `.get()` too?
The tests can be passed with this change:
`reserved_name = qs1.difference(qs2).values('name', 'order', 'id')[0]`

**Comment by felixxm** [comment]
`get()` works with combined queries so we shouldn't raise an exception in this case.

File: django/db/models/query.py [hunk 5]
```diff
@@ -338,6 +338,20 @@ def __or__(self, other):
...
+
+    def _supported_operation(func):
+        """Decorator function to check supported operation after combinators"""
+        @wraps(func)
+        def wrapped(self, *args, **kwargs):
+            if self.query.combinator:
+                raise NotSupportedError('Calling %s() is not supported after %s().' %
+                                        (func.__name__, self.query.combinator))
+            return func(self, *args, **kwargs)
+        return wrapped
```
Comment (felixxm) [inline django/db/models/query.py]: Sorry for the previous suggestion but decorator looks weird to me :smile: Method would be better, e.g.
```python
def _not_support_combined_queries(self, name):
    if self.query.combinator:
        raise NotSupportedError(
            'Calling %s() is not supported after %s().' % (name, self.query.combinator)
        )
```

**Comment by felixxm** [comment]
I found quite confusing the name of `Query.can_filter()` because we use it in many context but always to check if query is sliced, I'm going to rename it in a separate PR to the `Query.has_limit` property.

File: django/db/models/query.py [hunk 6]
```diff
@@ -885,18 +886,28 @@ def all(self):
...
+            )
+
     def filter(self, *args, **kwargs):
         """
         Return a new QuerySet instance with the args ANDed to the existing
         set.
         """
+        # Ignore check if filter called by .get()
+        if inspect.stack()[1].function != 'get':
+            self._not_support_combined_queries('filter')
```
Comment (felixxm) [inline django/db/models/query.py]: It's a bit :scream: to me. `filter()` doesn't change anything for combined queries so I think we should handle this in `get()`, e.g.
```python
def get(self, *args, **kwargs):
    clone = self._chain() if self.query.combinator else self.filter(*args, **kwargs)
```

Comment (hramezani) [reply django/db/models/query.py]: Thanks

File: tests/queries/test_qs_combinators.py [hunk 7]
```diff
@@ -258,3 +258,13 @@ def test_order_by_same_type(self):
         numbers = list(range(10))
         self.assertNumbersEqual(union.order_by('num'), numbers)
         self.assertNumbersEqual(union.order_by('other_num'), reversed(numbers))
+
+    def test_unsupported_operations(self):
+        qs = Number.objects.all()
+        msg = 'Calling %s() is not supported after %s().'
+        for combinator in ('union', 'intersection', 'difference'):
+            for method in ('filter', 'exclude', 'defer', 'only', 'extra',
+                           'prefetch_related', 'select_related'):
```
Comment (felixxm) [inline tests/queries/test_qs_combinators.py]: We should add `annotate()` to the list, also please use hanging indentation.

Comment (hramezani) [reply tests/queries/test_qs_combinators.py]: `annotate()` added.

**Comment by felixxm** [comment]
@hramezani Thanks :+1: I pushed minor edits.

**Pull request merged**